### DISCUSSION TITLE
Flow Formula

### DISCUSSION THREAD

**Discussion by AviadorLP** [ulic75/power-flow-card#27] [Ideas]
Maybe adding a user defined value for total flow as a fixed number.
(if its not defined or 0 let you current formula run)
My house can deal with 10kW and my inverted can deal only with 5kW
I probably would use 7500W for myself, but each user has its own limitations. 
So if anywhere some flow is higher then 7500 it would show with max flow speed,
and with this formula all low power flows will be always slow even if there are only low power flows
(i guess that is what people want)

**Comment by AviadorLP** [comment]
So in short what i'm saying is try to use the formula:

max - (value / "user_defined_total") * (max - min)

I think

although if value > "user_defined_total" then value should equal "user_defined_total"
in other words if value > "user_defined_total" then formula should simply be min ?????

i really don't understand your formula
is value the power flow value ?
is total all the flows added together ?
max and min the users defined max_flow_rate and min_flow_rate options ???

if so then if value equals total you get min, and if value equals 0 you get max...
should it not be exactly the opposite ?
Unless you are using this as an inverted flow speed, then i guess it would work.

**Comment by AviadorLP** [comment]
Anyway i believe you can just and very easily allow a "user define total" that replaces your var total if defined or not 0.
The speed will be your max if value reaches "user define total" and will even be greater if it exceeds it.
Try it i think it will work great and if the users prefer the way it is they just let it undefined....

**Comment by ulic75** [comment]
It's essentially the same formula as used by official HA Energy Distribution card.

- **value:** the given value of the flow, for example, 3kW grid to home
- **total:** all current flow values added together
- **max:** user defined max_flow_rate (defaulting to 6)
- **min:** user defined min_flow_rate (defaulting to 0.75)

The result being the amount of seconds it takes for a dot to traverse the path.  So the lower the number the faster it moves.

The "issue" that I'd like to fix is that if only one flow is currently active the speed will always be the same.  Grid to Home at 3kW would move at the same speed as Grid to Home at 30kW.

**Reply by AviadorLP** [reply]
I understand, and what i said will solve this, because the problem is total: all current flow values added together
it makes any value that is equal to total, work at max flow speed. Just try what i said. It will work!
replace total with a user defined value (just could not see in your code if this is a value in kW or in W)
And its a very easy fix.

Without any programming you can try this and check for yourself
change line 56 of power-flow-card.ts

return max - (value / total) * (max - min); 

with a fixed number for total (ex 10kW or 10000W or a lower if speeds are too slow,
 or adjust a higher min flow rate)
(i dont know if your total is dealing with kW or with W that is why i am giving you both values)
should not exceed your max grid capacity

return max - (value / 10) * (max - min);
or with
return max - (value / 10000) * (max - min);

And just watch the result....
i think that is where you calculate the speed, if i am not mistaken about this it will work

**Reply by ulic75** [reply]
We'd need to wrap it in a Math.max(max - (value / 10000) * (max - min), min) to ensure that it never goes below min (more importantly never goes below 0.001s).  This does seem to be an improvement assuming a comfortable user defined "total" is used.  I'm not certain I'm ready to make a change just yet though.

Here is a [spreadsheet](https://docs.google.com/spreadsheets/d/11TYPr3mKp1eRlrQgQdM8LjwqgXeBjdnl8G3zzzpbvig/edit?usp=sharing) I've used for playing with formulas.

**Reply by AviadorLP** [reply]
well, if you implement this, the good thing is a user can choose to keep the old formula or to define everything to his desire. If a user is careful choosing the user defined total and the min and the max. He can completely get the flows to his exact taste. I think you should since nothing will change to users that don't like this. They just let this var "total" undefined.
And this is what users have been asking for. In the readme me you should advice for a value between the inverter max power capability and the grid max power capability. I think that should be what people should be aiming for...

**Reply by AviadorLP** [reply]
And in the readme you should explain that this "total" is the value at witch the flow rate reaches its max.
Easy for anyone to understand.

**Comment by AviadorLP** [comment]
I believe your proposal #2  to wrap it in a Math.max(max - (value / user_total) * (max - min), min)
is the best way to go.

Also never a good idea for this value being less than you possible maximum solar production
In your spreadsheet in the example you have a solar productions of 16500 and choose a value of 7500.
That was a strange value you should choose a higher value and if required set higher min and max flow rates.
Otherwise all flows higher than 7500 will have the same speed and you have one with 10000w
